File: qa/project-browser-check.py
"""Browser acceptance against the production and standalone builds.
Run: python qa/project-browser-check.py http://127.0.0.1:4173
Controls receive real pointer events; no production test flags or fake results.
"""
from pathlib import Path
import json
import sys
import os
from playwright.sync_api import sync_playwright

BASE = sys.argv[1] if len(sys.argv) > 1 else 'http://127.0.0.1:4173'
ROOT = Path(__file__).resolve().parents[1]
OUT = ROOT / 'qa-results'
OUT.mkdir(exist_ok=True)
checks = []
errors = []
console = []
capture_warnings = []


def check(ok, label):
    assert ok, label
    checks.append(label)
    (OUT / 'browser-progress.json').write_text(json.dumps({'checks': checks}, indent=2))
    print('PASS:', label, flush=True)


def ready(page):
    page.wait_for_function('window.wiringLab && window.wiringLab.getProject', polling=100, timeout=30000)


def load(page, filename):
    page.locator('[data-project-file]').set_input_files(ROOT / 'examples' / filename)
    page.wait_for_function("document.querySelector('[data-project-status]').textContent.includes('已匯入')", polling=100, timeout=90000)
    assert '模擬未執行' in page.locator('[data-project-status]').inner_text()


def energized(page, component, value):
    page.wait_for_function("([id,value])=>window.wiringLab.getSimulation().result?.evaluation?.loads.some(x=>x.component===id && (x.state==='energized')===value)", arg=[component, value], polling=100, timeout=10000)


def pointer_target(button):
    # Independent DOM geometry/hit checks retain real pointer semantics without
    # requiring a stable WebGL animation frame for a static HTML control.
    box = button.evaluate('''el=>{el.scrollIntoView({block:'center',behavior:'instant'});
      const r=el.getBoundingClientRect(),x=r.x+r.width/2,y=r.y+r.height/2;
      return {x:r.x,y:r.y,width:r.width,height:r.height,enabled:!el.disabled,
        hit:el.contains(document.elementFromPoint(x,y))};}''')
    assert box['enabled'] and box['hit'] and box['width'] > 0 and box['height'] > 0, box
    return box['x'] + box['width']/2, box['y'] + box['height']/2


def click(page, button):
    page.mouse.click(*pointer_target(button))


def toggle_qf(page):
    page.locator('#component-select').select_option('QF1')
    click(page, page.locator('#details .actions button').filter(has_text='切換 ON／OFF'))


def press(page, component, assertion):
    page.locator('#component-select').select_option(component)
    button = page.locator('#details .actions button').filter(has_text='按住手動壓合')
    page.mouse.move(*pointer_target(button))
    page.mouse.down()
    try:
        assertion()
    finally:
        page.mouse.up()


def capture(page, name):
    # Capturing a software-rendered frame must not prevent the remaining
    # functional checks from running. A timeout is recorded, never called a pass.
    try:
        page.screenshot(path=str(OUT / name), full_page=False, timeout=10000)
    except Exception as error:
        capture_warnings.append({'file': name, 'error': str(error)})
        print('CAPTURE WARNING:', name, str(error), flush=True)
        (OUT / 'capture-warnings.json').write_text(json.dumps(capture_warnings, indent=2))


def run(browser, page):
    page.goto(BASE, wait_until='networkidle')
    ready(page)
    check(page.evaluate('window.wiringLab.getRevision()') == 'WIRE-R17', 'R17 boot')
    check(page.locator('#component-select option').count() == 22, 'default 22 components')
    load(page, 'a04-motor-start.project.json')
    check(page.evaluate('window.wiringLab.getWires().length') == 28, 'A04 native route import')
    check(page.locator('[data-source], .supply-switches').count() == 0, 'no independent source controls')
    check(page.locator('.equipment-card').count() == 2 and page.locator('[data-equipment="CONTROL"]').count() == 0, 'only MAIN supply and M1 motor equipment')
    check(not page.evaluate("window.wiringLab.getProject().configuration.components.find(c=>c.id==='QF1').state.on"), 'A04 defaults to QF1 OFF')
    click(page, page.locator('[data-sim-start]'))
    energized(page, 'MC1', False)
    press(page, 'PB3', lambda: energized(page, 'MC1', False))
    toggle_qf(page)
    press(page, 'PB3', lambda: energized(page, 'MC1', True))
    energized(page, 'MC1', True)
    page.wait_for_function("window.wiringLab.getSimulation().result.evaluation.motors.find(m=>m.component==='M1').state==='powered'", polling=100)
    toggle_qf(page)
    energized(page, 'MC1', False)
    energized(page, 'HL4', False)
    toggle_qf(page)
    energized(page, 'MC1', False)
    check(True, 'QF1 cuts coil and lamp, restoration with released ON does not restart')
    press(page, 'PB3', lambda: energized(page, 'MC1', True))
    check(True, 'start, release, self-hold and three-phase motor')
    press(page, 'PB5', lambda: energized(page, 'MC1', False))
    check(True, 'stop button')
    press(page, 'PB3', lambda: energized(page, 'MC1', True))
    page.locator('#component-select').select_option('TH1')
    click(page, page.locator('#details button').filter(has_text='TEST'))
    energized(page, 'MC1', False)
    energized(page, 'HL3', True)
    energized(page, 'BZ1', True)
    check(True, 'overload, red lamp and buzzer')
    toggle_qf(page)
    energized(page, 'HL3', False)
    energized(page, 'BZ1', False)
    toggle_qf(page)
    energized(page, 'HL3', True)
    energized(page, 'BZ1', True)
    check(True, 'QF1 also removes and restores the real overload-alarm supply')
    page.locator('#component-select').select_option('TH1')
    click(page, page.locator('#details button').filter(has_text='RESET'))
    energized(page, 'MC1', False)
    energized(page, 'HL3', False)
    check(True, 'RESET does not restart')
    assert len(page.evaluate('window.wiringLab.getSimulation().evaluatedCircuit.threePhaseSources')) == 1
    check(page.evaluate('window.wiringLab.getSimulation().evaluatedCircuit.sources.length') == 0, 'actual solver has one original source, no CONTROL')
    capture(page, 'a04-loaded.png')
    click(page, page.locator('[data-sim-stop]'))
    for _ in range(2):
        click(page, page.locator('#flap-btn'))
        assert page.evaluate('window.wiringLab.getProject().configuration.operationPanel.state.open')
        click(page, page.locator('#flap-btn'))
        assert not page.evaluate('window.wiringLab.getProject().configuration.operationPanel.state.open')
    check(True, 'two panel open-close cycles')
    with page.expect_download() as download:
        click(page, page.locator('[data-project-export]'))
    target = OUT / 'browser-export.project.json'
    download.value.save_as(target)
    exported = json.loads(target.read_text())
    check(exported['format'] == 'wiring-panel-project' and len(exported['connections']) == 31 and all(set(w) == {'from', 'to'} for w in exported['connections']), 'download contains endpoints only')
    load(page, 'custom-panel-less.project.json')
    check(page.locator('#component-select option').count() == 2 and page.locator('.equipment-card').count() == 0 and page.locator('#flap-btn').is_hidden(), 'different layout replaces old project')
    page.locator('#component-select').select_option('coil')
    click(page, page.locator('#inspect-component'))
    assert '返回' in page.locator('#inspect-component').inner_text()
    click(page, page.locator('#inspect-component'))
    check(True, 'generic component inspection')
    before = page.evaluate('window.wiringLab.getProject()')
    page.locator('[data-project-file]').set_input_files(ROOT / 'examples/a04-motor-start.project.json')
    page.wait_for_function("document.querySelector('[data-project-status]').textContent.includes('自動走線')", polling=100, timeout=30000)
    click(page, page.locator('[data-project-cancel]'))
    page.wait_for_function("document.querySelector('[data-project-status]').textContent.includes('取消') && !document.querySelector('[data-project-import]').disabled", polling=100, timeout=30000)
    check(page.evaluate('window.wiringLab.getProject()') == before, 'cancel preserves previous project')
    load(page, 'shared-source-drives.project.json')
    assert page.locator('.equipment-card').count() == 1
    click(page, page.locator('[data-sim-start]'))
    energized(page, 'coil', True)
    energized(page, 'coil-B', False)
    page.locator('#component-select').select_option('thermal-B')
    click(page, page.locator('#details button').filter(has_text='RESET'))
    energized(page, 'coil-B', True)
    page.locator('#component-select').select_option('thermal')
    click(page, page.locator('#details button').filter(has_text='TEST'))
    energized(page, 'coil', False)
    energized(page, 'coil-B', True)
    check(True, 'one shared source, two independent physical thermal contacts')
    click(page, page.locator('[data-sim-stop]'))
    click(page, page.locator('#reset-project'))
    page.wait_for_function("window.wiringLab.getProject().name==='BOARD 024' && document.querySelector('#component-select').options.length===22", polling=100)
    check(page.evaluate('window.wiringLab.getWires().length') == 0, 'reset to default')
    page.close()
    # Opening through both HTTP and file:// checks the delivered offline artifact.
    for url, label in [(BASE + '/wiring-panel.html', 'served standalone'),
                       ((ROOT / 'downloads/wiring-panel-WIRE-R17.html').as_uri(), 'file-protocol standalone')]:
        offline = browser.new_page(viewport={'width': 1500, 'height': 1000})
        offline.on('pageerror', lambda error: errors.append(str(error)))
        offline.goto(url, wait_until='load')
        ready(offline)
        check(offline.evaluate("window.wiringLab.getRevision()==='WIRE-R17' && window.wiringLab.getProject().format==='wiring-panel-project'"), label + ' boot')
        load(offline, 'custom-panel-less.project.json')
        check(offline.locator('#component-select option').count() == 2, label + ' project import')
        offline.close()
    check(not errors, 'no uncaught browser exceptions')


with sync_playwright() as playwright:
    browser = playwright.chromium.launch(headless=True, executable_path=os.environ.get('CHROMIUM_EXECUTABLE'), args=['--use-angle=swiftshader', '--enable-unsafe-swiftshader'])
    page = browser.new_page(viewport={'width': 1500, 'height': 1000}, accept_downloads=True)
    page.on('pageerror', lambda error: errors.append(str(error)))
    page.on('console', lambda message: console.append(message.type + ': ' + message.text))
    try:
        run(browser, page)
    except Exception as error:
        details = {'checks': checks, 'error': str(error), 'pageErrors': errors, 'console': console[-50:]}
        if not page.is_closed():
            try:
                details['ui'] = page.evaluate("({status:document.querySelector('[data-project-status]')?.textContent, selection:document.querySelector('#component-select')?.value, inspector:document.querySelector('#details')?.innerText})")
                capture(page, 'browser-failure.png')
            except Exception as capture_error:
                details['captureError'] = str(capture_error)
        (OUT / 'browser-failure.json').write_text(json.dumps(details, ensure_ascii=False, indent=2))
        raise
    finally:
        browser.close()
(OUT / 'browser-results.json').write_text(json.dumps({'status': 'passed', 'checks': checks, 'pageErrors': errors, 'captureWarnings': capture_warnings}, ensure_ascii=False, indent=2))
print('PASS: browser acceptance and standalone entries', flush=True)

File: qa/project-browser-check.ts
/** Optional real-pointer acceptance for production and standalone builds.
 * Run after npm run build and serving dist:
 * npm run test:browser:project -- http://127.0.0.1:4173
 * Requires npx playwright install chromium, or CHROMIUM_EXECUTABLE.
 * Not part of npm test: this script creates a WebGL renderer.
 */
import assert from 'node:assert/strict';
import {mkdirSync, readFileSync, writeFileSync} from 'node:fs';
import {dirname, join, resolve} from 'node:path';
import {fileURLToPath, pathToFileURL} from 'node:url';
import {chromium, type Browser, type Locator, type Page} from 'playwright';
import {MODEL_REVISION} from '../src/revision.ts';
import type {ProjectDocument} from '../src/project/contracts.ts';

const BASE = (process.argv[2] ?? 'http://127.0.0.1:4173').replace(/\/$/, '');
const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const OUT = join(ROOT, 'qa-results');
const checks: string[] = [], errors: string[] = [], messages: string[] = [];
const captureWarnings: {file: string; error: string}[] = [];
const messageOf = (error: unknown) => error instanceof Error ? error.message : String(error);
function writeJSON(name: string, value: unknown): void {
  writeFileSync(join(OUT, name), JSON.stringify(value, null, 2) + '\n');
}
function check(ok: unknown, label: string): void {
  assert.ok(ok, label);
  checks.push(label);
  writeJSON('browser-progress.json', {checks});
  console.log('PASS:', label);
}
async function ready(page: Page): Promise<void> {
  await page.waitForFunction(() => window.wiringLab && 'getProject' in window.wiringLab, undefined, {polling: 100, timeout: 30000});
  // Labels are assertions too; make the QA locale explicit without changing app defaults.
  await page.evaluate(() => window.wiringLab?.setLocale?.('zh-TW'));
}
async function state(page: Page) {
  return page.evaluate(() => {
    const lab = window.wiringLab;
    if (!lab || !('getProject' in lab)) throw Error('Project app did not initialize');
    return {revision: lab.getRevision(), project: lab.getProject(), simulation: lab.getSimulation(), wires: lab.getWires()};
  });
}
async function load(page: Page, filename: string): Promise<void> {
  await page.locator('[data-project-file]').setInputFiles(join(ROOT, 'examples', filename));
  await page.waitForFunction(() => document.querySelector('[data-project-status]')?.textContent?.includes('已匯入'), undefined, {polling: 100, timeout: 90000});
  assert.match(await page.locator('[data-project-status]').innerText(), /模擬未執行/);
}
async function energized(page: Page, component: string, value: boolean): Promise<void> {
  await page.waitForFunction(([id, expected]) => {
    const lab = window.wiringLab;
    return lab && 'getProject' in lab && lab.getSimulation().result?.evaluation?.loads.some(load => load.component === id && (load.state === 'energized') === expected);
  }, [component, value] as const, {polling: 100, timeout: 10000});
}
async function pointerTarget(button: Locator): Promise<[number, number]> {
  // Check the actual DOM hit target, then send real pointer events.
  const box = await button.evaluate(el => {
    el.scrollIntoView({block: 'center', behavior: 'instant'});
    const r = el.getBoundingClientRect(), x = r.x + r.width / 2, y = r.y + r.height / 2;
    return {x: r.x, y: r.y, width: r.width, height: r.height, enabled: el instanceof HTMLButtonElement && !el.disabled, hit: el.contains(document.elementFromPoint(x, y))};
  });
  assert.ok(box.enabled && box.hit && box.width > 0 && box.height > 0, JSON.stringify(box));
  return [box.x + box.width / 2, box.y + box.height / 2];
}
async function click(page: Page, button: Locator): Promise<void> {
  await page.mouse.click(...await pointerTarget(button));
}
async function toggleQF(page: Page): Promise<void> {
  await page.locator('#component-select').selectOption('QF1');
  await click(page, page.locator('#details .actions button').filter({hasText: '切換 ON／OFF'}));
}
async function press(page: Page, component: string, assertion: () => Promise<void>): Promise<void> {
  await page.locator('#component-select').selectOption(component);
  const button = page.locator('#details .actions button').filter({hasText: '按住手動壓合'});
  await page.mouse.move(...await pointerTarget(button));
  await page.mouse.down();
  try {await assertion();} finally {await page.mouse.up();}
}
async function capture(page: Page, name: string): Promise<void> {
  // A software-render screenshot timeout is recorded as a warning, never a pass.
  try {await page.screenshot({path: join(OUT, name), fullPage: false, timeout: 10000});}
  catch (error) {
    captureWarnings.push({file: name, error: messageOf(error)});
    console.warn('CAPTURE WARNING:', name, messageOf(error));
    writeJSON('capture-warnings.json', captureWarnings);
  }
}
async function run(browser: Browser, page: Page): Promise<void> {
  await page.goto(BASE, {waitUntil: 'networkidle'});
  await ready(page);
  check((await state(page)).revision === MODEL_REVISION, `${MODEL_REVISION} boot`);
  check(await page.locator('#component-select option').count() === 22, 'default 22 components');
  await load(page, 'a04-motor-start.project.json');
  check((await state(page)).wires.length === 28, 'A04 native route import');
  check(await page.locator('[data-source], .supply-switches').count() === 0, 'no independent source controls');
  check(await page.locator('.equipment-card').count() === 2 && await page.locator('[data-equipment="CONTROL"]').count() === 0, 'only MAIN supply and M1 motor equipment');
  const breaker = (await state(page)).project.configuration.components.find(c => c.id === 'QF1');
  assert.ok(breaker?.state);
  check(!breaker.state.on, 'A04 defaults to QF1 OFF');
  await click(page, page.locator('[data-sim-start]'));
  await energized(page, 'MC1', false);
  await press(page, 'PB3', () => energized(page, 'MC1', false));
  await toggleQF(page);
  await press(page, 'PB3', () => energized(page, 'MC1', true));
  await energized(page, 'MC1', true);
  await page.waitForFunction(() => {
    const lab = window.wiringLab;
    return lab && 'getProject' in lab && lab.getSimulation().result?.evaluation?.motors.find(m => m.component === 'M1')?.state === 'powered';
  }, undefined, {polling: 100});
  await toggleQF(page);
  await energized(page, 'MC1', false);
  await energized(page, 'HL4', false);
  await toggleQF(page);
  await energized(page, 'MC1', false);
  check(true, 'QF1 cuts coil and lamp, restoration with released ON does not restart');
  await press(page, 'PB3', () => energized(page, 'MC1', true));
  check(true, 'start, release, self-hold and three-phase motor');
  await press(page, 'PB5', () => energized(page, 'MC1', false));
  check(true, 'stop button');
  await press(page, 'PB3', () => energized(page, 'MC1', true));
  await page.locator('#component-select').selectOption('TH1');
  await click(page, page.locator('#details button').filter({hasText: 'TEST'}));
  await energized(page, 'MC1', false);
  await energized(page, 'HL3', true);
  await energized(page, 'BZ1', true);
  check(true, 'overload, red lamp and buzzer');
  await toggleQF(page);
  await energized(page, 'HL3', false);
  await energized(page, 'BZ1', false);
  await toggleQF(page);
  await energized(page, 'HL3', true);
  await energized(page, 'BZ1', true);
  check(true, 'QF1 also removes and restores the real overload-alarm supply');
  await page.locator('#component-select').selectOption('TH1');
  await click(page, page.locator('#details button').filter({hasText: 'RESET'}));
  await energized(page, 'MC1', false);
  await energized(page, 'HL3', false);
  check(true, 'RESET does not restart');
  const evaluated = (await state(page)).simulation.evaluatedCircuit;
  assert.ok(evaluated);
  assert.equal(evaluated.threePhaseSources?.length, 1);
  check(evaluated.sources.length === 0, 'actual solver has one original source, no CONTROL');
  await capture(page, 'a04-loaded.png');
  await click(page, page.locator('[data-sim-stop]'));
  for (let i = 0; i < 2; i++) {
    await click(page, page.locator('#flap-btn'));
    assert.equal((await state(page)).project.configuration.operationPanel?.state.open, true);
    await click(page, page.locator('#flap-btn'));
    assert.equal((await state(page)).project.configuration.operationPanel?.state.open, false);
  }
  check(true, 'two panel open-close cycles');
  const downloading = page.waitForEvent('download');
  await click(page, page.locator('[data-project-export]'));
  const target = join(OUT, 'browser-export.project.json');
  await (await downloading).saveAs(target);
  const exported: ProjectDocument = JSON.parse(readFileSync(target, 'utf8'));
  check(exported.format === 'wiring-panel-project' && exported.connections.length === 31 && exported.connections.every(w => JSON.stringify(Object.keys(w).sort()) === '["from","to"]'), 'download contains endpoints only');
  await load(page, 'custom-panel-less.project.json');
  check(await page.locator('#component-select option').count() === 2 && await page.locator('.equipment-card').count() === 0 && await page.locator('#flap-btn').isHidden(), 'different layout replaces old project');
  await page.locator('#component-select').selectOption('coil');
  await click(page, page.locator('#inspect-component'));
  assert.match(await page.locator('#inspect-component').innerText(), /返回/);
  await click(page, page.locator('#inspect-component'));
  check(true, 'generic component inspection');
  const before = (await state(page)).project;
  await page.locator('[data-project-file]').setInputFiles(join(ROOT, 'examples/a04-motor-start.project.json'));
  await page.waitForFunction(() => document.querySelector('[data-project-status]')?.textContent?.includes('自動走線'), undefined, {polling: 100, timeout: 30000});
  await click(page, page.locator('[data-project-cancel]'));
  await page.waitForFunction(() => document.querySelector('[data-project-status]')?.textContent?.includes('取消') && !document.querySelector<HTMLButtonElement>('[data-project-import]')?.disabled, undefined, {polling: 100, timeout: 30000});
  assert.deepEqual((await state(page)).project, before);
  check(true, 'cancel preserves previous project');
  await load(page, 'shared-source-drives.project.json');
  assert.equal(await page.locator('.equipment-card').count(), 1);
  await click(page, page.locator('[data-sim-start]'));
  await energized(page, 'coil', true);
  await energized(page, 'coil-B', false);
  await page.locator('#component-select').selectOption('thermal-B');
  await click(page, page.locator('#details button').filter({hasText: 'RESET'}));
  await energized(page, 'coil-B', true);
  await page.locator('#component-select').selectOption('thermal');
  await click(page, page.locator('#details button').filter({hasText: 'TEST'}));
  await energized(page, 'coil', false);
  await energized(page, 'coil-B', true);
  check(true, 'one shared source, two independent physical thermal contacts');
  await click(page, page.locator('[data-sim-stop]'));
  await click(page, page.locator('#reset-project'));
  await page.waitForFunction(() => {
    const lab = window.wiringLab;
    return lab && 'getProject' in lab && lab.getProject().name === 'BOARD 024' && document.querySelector<HTMLSelectElement>('#component-select')?.options.length === 22;
  }, undefined, {polling: 100});
  check((await state(page)).wires.length === 0, 'reset to default');
  await page.close();
  for (const [url, label] of [[BASE + '/wiring-panel.html', 'served standalone'], [pathToFileURL(join(ROOT, `downloads/wiring-panel-${MODEL_REVISION}.html`)).href, 'file-protocol standalone']]) {
    const offline = await browser.newPage({locale: 'zh-TW', viewport: {width: 1500, height: 1000}});
    offline.on('pageerror', error => errors.push(messageOf(error)));
    try {
      await offline.goto(url, {waitUntil: 'load'});
      await ready(offline);
      const boot = await state(offline);
      check(boot.revision === MODEL_REVISION && boot.project.format === 'wiring-panel-project', label + ' boot');
      await load(offline, 'custom-panel-less.project.json');
      check(await offline.locator('#component-select option').count() === 2, label + ' project import');
    } finally {await offline.close();}
  }
  check(errors.length === 0, 'no uncaught browser exceptions');
}
async function main(): Promise<void> {
  if (BASE === '--help') {
    console.log('Usage: npm run test:browser:project -- [production-base-url]');
    return;
  }
  mkdirSync(OUT, {recursive: true});
  const browser = await chromium.launch({headless: true, executablePath: process.env.CHROMIUM_EXECUTABLE, args: ['--use-angle=swiftshader', '--enable-unsafe-swiftshader']});
  const page = await browser.newPage({locale: 'zh-TW', viewport: {width: 1500, height: 1000}, acceptDownloads: true});
  page.on('pageerror', error => errors.push(messageOf(error)));
  page.on('console', message => messages.push(message.type() + ': ' + message.text()));
  try {
    await run(browser, page);
  } catch (error) {
    const details: Record<string, unknown> = {checks, error: messageOf(error), pageErrors: errors, console: messages.slice(-50)};
    if (!page.isClosed()) {
      try {
        details.ui = await page.evaluate(() => ({status: document.querySelector('[data-project-status]')?.textContent, selection: document.querySelector<HTMLSelectElement>('#component-select')?.value, inspector: document.querySelector<HTMLElement>('#details')?.innerText}));
        await capture(page, 'browser-failure.png');
      } catch (captureError) {details.captureError = messageOf(captureError);}
    }
    writeJSON('browser-failure.json', details);
    throw error;
  } finally {await browser.close();}
  writeJSON('browser-results.json', {status: 'passed', checks, pageErrors: errors, captureWarnings});
  console.log('PASS: browser acceptance and standalone entries');
}
await main();
